Spigot-API-Patches/0189-Add-Mob-Goal-API.patch

```diff
@@ -0,0 +1,341 @@
+From a80ab2bc1135e7a6ab3d2c56aa7ddeb4168b954a Mon Sep 17 00:00:00 2001
+From: MiniDigger <admin@minidigger.me>
+Date: Fri, 3 Jan 2020 16:24:46 +0100
+Subject: [PATCH] Add Mob Goal API
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/ai/Goal.java b/src/main/java/com/destroystokyo/paper/ai/Goal.java
+new file mode 100644
+index 00000000..340055ed
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/ai/Goal.java
+@@ -0,0 +1,54 @@
++package com.destroystokyo.paper.ai;
++
++import org.jetbrains.annotations.NotNull;
++
++import org.bukkit.NamespacedKey;
++
++/**
++ * Represents an AI goal of an entity
++ */
++public interface Goal {
++
++    /**
++     * Checks if this goal should be activated
++     *
++     * @return if this goal should be activated
++     */
++    boolean shouldActivate();
++
++    /**
++     * Checks if this goal should stay active, defaults to {@link Goal#shouldActivate()}
++     *
++     * @return if this goal should stay active
++     */
++    default boolean shouldStayActive() {
++        return shouldActivate();
++    }
++
++    /**
++     * Called when this goal gets activated
++     */
++    default void start() {
++    }
++
++    /**
++     * Called when this goal gets stopped
++     */
++    default void stop() {
++    }
++
++    /**
++     * Called each tick the goal is activated
++     */
++    default void tick() {
++    }
++
++    /**
++     * A unique key that identifies this type of goal. Plugins should use their own namespace, not the minecraft
++     * namespace
++     *
++     * @return a namespaced key.
++     */
++    @NotNull
++    NamespacedKey getKey();
++}
+diff --git a/src/main/java/com/destroystokyo/paper/ai/GoalType.java b/src/main/java/com/destroystokyo/paper/ai/GoalType.java
+new file mode 100644
+index 00000000..fbda833d
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/ai/GoalType.java
+@@ -0,0 +1,14 @@
++package com.destroystokyo.paper.ai;
++
++import org.jetbrains.annotations.NotNull;
++
++import org.bukkit.NamespacedKey;
++
++public interface GoalType {
++
++    @NotNull
++    NamespacedKey getKey();
++
++    GoalType TARGET = () -> new NamespacedKey("paper", "target");
++    GoalType BEHAVIOR = () -> new NamespacedKey("paper", "goal");
++}
```

Comment:
Seeing as this doesn't seem to be meant to be implemented by plugins to create new types it would be good to at least mention that in the docs (or just move to an Enum, but that might break future compatibility if Mojang starts to support custom types)

Reply:
I remember that MC itself has no enum or something else.
They've target- and goalSelector attributes. We discussed that and came to the conclusion, that an enum like solution is better regarding later expandability.
Converting it to an enum or properly documenting it would fix the confusion. 
An alternative to keep future compatibility would be to follow an approach how GameRule is implemented. We pass a/n class/interface as a generic parameter.

```java

public final class GoalType<T> {

    private NamespacedKey key;
    private Class<T> clazz;
    @NotNull
    Class<T> getInterface(){
      return clazz;
    }
    @NotNull
    NamespacedKey getKey(){
      return key;
   }
   private GoalType(Class<T> clazz, NamespacedKey key){
       this.clazz = clazz;
       this.key = key;
   }
   

    public static final GoalType<Goal> TARGET = new GoalType(Goal.class, new NamespacedKey("paper", "target"));
    public static final GoalType<Goal> BEHAVIOR = new GoalType(Goal.class, new NamespacedKey("paper", "goal"));
}
```

Additional the add/remove Goal needs to be generic as well.

```java
<T> void addGoal(@NotNull GoalType<T> type, int priority, @NotNull T goal);
<T> void removeGoal(@NotNull GoalType<T> type, @NotNull T goal);
```

![Screenshot 2020-01-14 at 18 27 38](https://user-images.githubusercontent.com/1420893/72367179-057da400-36fc-11ea-9a10-a1df2ebc4702.png)

Reply:
addressed this